### PR TITLE
import inital version of python styleguide

### PR DESCRIPTION
**Pull request by Nicoretti**
- Add README
- Add project configuration and dependencies
- Add .gitignore
- Add inital styleguide
- Add github action to generate github page(s)

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,32 @@
+Exasol python-styleguide
+========================
+Welcome to the Exasol_ python-styleguide.
+The styleguide(s) is not intend to be a universal text on how to do each single detail in python,
+it rather tries to reley on existing documents and guides references them (e.g. pep8_) and their rules or it
+will add useful additions or exceptions where Exasol_ is deverging from the *"norm"* and why.
```
Comment (tkilias) [inline README.rst]:
```suggestion
will add useful additions or exceptions where Exasol_ is diverging from the *"norm"* and why.
```

File: noxfile.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+BUILD_DIRECTORY = STYLEGUIDE / "_build"
+
+nox.options.sessions = ['build', 'open']
+
+
+@nox.session(python=None)
+def build(session):
+    """Build the styleguide."""
+    session.run(
+        "sphinx-build",
```
Comment (tkilias) [inline noxfile.py:18]: maybe introduce directly the sphinx github pages generator

File: noxfile.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            f"""
+            File {index_page} does not exist.
+            Please run `nox -s build` first
+            """
+            )
+        )
+    session.run(
+        "python",
+        "-m",
+        "webbrowser",
```
Comment (tkilias) [inline noxfile.py:49]: interesting

File: styleguide/conf.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+# directories to ignore when looking for source files.
+# This pattern also affects html_static_path and html_extra_path.
+exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
+
+# -- Options for HTML output -------------------------------------------------
+
+# The theme to use for HTML and HTML Help pages.  See the documentation for
+# a list of builtin themes.
+#
+html_theme = "furo"
```
Comment (tkilias) [inline styleguide/conf.py:52]: which theme we use we need to decice, I think in bucketfs-utils-python we the default one at the moment

File: styleguide/drafts/tooling.rst [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+Code Style
+++++++++++
+
+* Black_     (style, code formatter)
+* Isort_     (style, import sorting)
+
+Testing
++++++++
+
+* Pytest_    (ut, testing)
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst]: ut?

File: styleguide/drafts/tooling.rst [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+* Coverage_  (coverage)
+* MyPy_      (linter, type checking)
+* Sonar_     (linter, code coverage, code health)
+
+
+Up For Consideration
+++++++++++++++++++++
+
+* `scriv <https://github.com/nedbat/scriv>`_ (changelog management)
+* `prysk <https://www.prysk.net/>`_ (Integration testing cli's and co.)
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:68]: need to take a look at

File: styleguide/drafts/tooling.rst [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+
+* Poetry_    (dependency management, virtual environment, build-tool, packaging)
+
+Code Health
++++++++++++
+
+* Pylint_    (linter, code analysis)
+* Coverage_  (coverage)
+* MyPy_      (linter, type checking)
+* Sonar_     (linter, code coverage, code health)
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:61]: what do you think, will pylin and sonar collide?

File: styleguide/drafts/tooling.rst [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+* `tox <https://tox.wiki/en/latest/>`_
+* `make <https://en.wikipedia.org/wiki/Make_(software)>`_
+* `cargo-make <https://sagiegurari.github.io/cargo-make/>`_
+* `just <https://github.com/casey/just>`_
+* `Ant <https://ant.apache.org/>`_
+* `Task <https://taskfile.dev/#/>`_
+* `Tusk <https://github.com/rliebz/tusk>`_
+* `Gulp <https://gulpjs.com/>`_
+* `Grunt <https://gruntjs.com/>`_
+
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:130]: We should mention https://github.com/nat-n/poethepoet, because we still use it in some projects

File: styleguide/drafts/tooling.rst [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+      -
+    * - Good & Powerful CLI
+      -
+    * - No manual dependency editing required
+      -
+
+Alternatives
+~~~~~~~~~~~~
+
+* `pipenv <https://pipenv.pypa.io/en/latest/>`_
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:474]: should we explain, why we don't use pipenv? I have two reasons. First, the project was more than a year dead and no maintainer, currently it has one, but that wasn't a good sign and second, the resolver was not as powerful.

File: styleguide/drafts/tooling.rst [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+      -
+    * - Compatible with sonar
+      -
+    * - Supports pyproject.toml
+      -
+
+Alternatives
+~~~~~~~~~~~~
+
+* `Flake8 <https://flake8.pycqa.org/en/latest/>`_
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:516]: what is the benefit of pylint to the alternatives

File: styleguide/drafts/tooling.rst [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+
+Alternatives
+~~~~~~~~~~~~
+
+* `pytype <https://github.com/google/pytype>`_ (google)
+* `pyright <https://github.com/Microsoft/pyright>`_ (microsoft)
+* `pyre-check <https://github.com/facebook/pyre-check>` (facebook, contains security checking too)
+* IDE built in e.g. PyCharm
+
+PreCommit_
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst]: I thought this is part of automation

**Review by tkilias — CHANGES_REQUESTED** [review]

File: noxfile.py (hunk 2 above)
Comment (Nicoretti) [reply noxfile.py:18]: I would say can be added later if needed, especially because I don't think there is a need to have different versions of the styleguide. The styleguide is in flux and the current master should reflect the current state of the art.

File: noxfile.py (hunk 3 above)
Comment (Nicoretti) [reply noxfile.py:49]: The python interpreter/modules ship with quite cool `cli-tools` here are two more cool ones:
* `python -m json.tool -h`
* `python -m zipfile -h`

usually you can find at the bottom of the documentation for a specific module, if it has some sort of `cli`/`commands`
e.g. https://docs.python.org/3/library/zipfile.html#command-line-interface

File: styleguide/conf.py (hunk 4 above)
Comment (Nicoretti) [reply styleguide/conf.py:52]: Definitely it would make sense if we use a common theme for all of our documents.
Here is a quick wrap up why I have chosen `Furo` (like it in general)

* Readable Font
* Good Page Structure (e.g. sidbar left: overview, sidebar right: current location in the document)
* Supports/Works well on mobile devices too
  ([Give it a try](https://black.readthedocs.io/en/stable/))
* Darkmode support 
* Autodetection for darkmode
* Lots of other python project's adopt this theme
  (I think lots of them because it is nice looking and very functional/good for documentation)

File: styleguide/drafts/tooling.rst (hunk 5 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst]: ut = unit-test(s)

File: styleguide/drafts/tooling.rst (hunk 7 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:61]: Sonar supports pylint reports see https://docs.sonarcloud.io/enriching/external-analyzer-reports/

File: styleguide/drafts/tooling.rst (hunk 9 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:474]: We definitely wanna do explain why we have taken which choice in the final guide.

File: styleguide/drafts/tooling.rst (hunk 10 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:516]: In the end most of them are quite the same and they have overlaping features.
I personally have the most experience with `pylint` and I also like the following features of it:
* "rating" feature which helps to improve the code base over time without imposing super strict restrictions. 
   (one can commit as long as overall code quality does not decrease, without disallowing "new warnings".
   e.g. if you fix 2 warnings and introduce a new one, you are good to commit it)
* Lots helpful warnings (e.g. dangerous default parameters `def func(a=[])`)
* Plugin/Extension point ("if custom checks etc. are needed")

File: styleguide/drafts/tooling.rst (hunk 11 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst]: Mostly yes, but it would be a "dev-toos" dependency that's why I have it included here.

TL;DR: I would say the hooks itself and what they do is the automation having a tool which helps us here is "tooling"
(it manages the commit hooks)

File: noxfile.py (hunk 2 above)
Comment (tkilias) [reply noxfile.py:18]: Or where you want to go. And, it is clear that we want to deploy the documentation on GitHub pages and this with the Sphinx GitHub pages Generator

File: styleguide/conf.py (hunk 4 above)
Comment (tkilias) [reply styleguide/conf.py:52]: I don't have a preference, so this one is looking good, so I am fine with it. Could you create a ticket for the bucketfs utils python, that we switch to this one

File: styleguide/drafts/tooling.rst (hunk 5 above)
Comment (tkilias) [reply styleguide/drafts/tooling.rst]: Please, only use abbreviations after you have introduced them.

File: styleguide/drafts/tooling.rst (hunk 10 above)
Comment (tkilias) [reply styleguide/drafts/tooling.rst:516]: Sounds good, I used also pylint so far, but didn't look too much into the other ones and didn't have the comparison

File: styleguide/drafts/tooling.rst (hunk 11 above)
Comment (tkilias) [reply styleguide/drafts/tooling.rst]: But wouldn't be nox also a dev-tool? So, I would mention it after nox.

File: noxfile.py (hunk 2 above)
Comment (Nicoretti) [reply noxfile.py:18]: For me personally GH pages would be preferable, but if this is not ideal or disallowed, I am open to suggestions.
@redcatbear from company policy point of view would GH-Pages be ok?

File: styleguide/drafts/tooling.rst (hunk 10 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:516]: so bullet point 4. flat learning curve for at least 2 python devs ;)

Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:516]: Note: 
I also never did a in detail comparison, but from my experience pylint does its job well.

File: styleguide/drafts/tooling.rst (hunk 11 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst]: Ok understood, makes sense I'll do.

File: noxfile.py (hunk 2 above)
Comment (tomuben) [reply noxfile.py:18]: @Nicoretti , yes it is. @MarleneKress79789 and @tkilias worked on 
https://github.com/exasol/sphinx-github-pages-generator
which simplifies the generation and deployment of GH pages.

I believe Torstens suggestion is to use this tool also here.

File: README.rst [hunk 12]
```diff
@@ -0,0 +1,32 @@
+Exasol python-styleguide
+========================
+Welcome to the Exasol_ python-styleguide.
+The styleguide(s) is not intend to be a universal text on how to do each single detail in python,
```
Comment (tomuben) [inline README.rst]:
```suggestion
The styleguide(s) is not intend to be an universal text on how to do each single detail in python,
```

File: README.rst [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+it rather tries to reley on existing documents and guides references them (e.g. pep8_) and their rules or it
```
Comment (tomuben) [inline README.rst]:
```suggestion
it rather tries to rely on existing documents and guides references them (e.g. pep8_) and their rules or it
```

File: README.rst [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+will add useful additions or exceptions where Exasol_ is diverging from the *"norm"* and why.
```
Comment (tomuben) [inline README.rst]:
```suggestion
will add useful additions or exceptions where Exasol_ is diverging from the *"norm"* and will explain the reasons behind.
```

File: pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+[tool.poetry]
+name = "python-styleguide"
```
Comment (tomuben) [inline pyproject.toml:2]:
```suggestion
name = "exasol-python-styleguide"
```

File: pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+version = "0.1.0"
+description = ""
+authors = ["Your Name <you@example.com>"]
```
Comment (tomuben) [inline pyproject.toml]: Missing...

File: styleguide/conf.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+# documentation root, use os.path.abspath to make it absolute, like shown here.
+#
+# import os
+# import sys
+# sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'Python Styleguide'
```
Comment (tomuben) [inline styleguide/conf.py:20]:
```suggestion
project = 'Exasol Python Styleguide'
```

File: styleguide/conf.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+copyright = '2022, Exasol'
+author = 'Exasol'
```
Comment (tomuben) [inline styleguide/conf.py]:
```suggestion
author = 'Exasol AG'
```

File: styleguide/conf.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+# This pattern also affects html_static_path and html_extra_path.
+exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
+
+# -- Options for HTML output -------------------------------------------------
+
+# The theme to use for HTML and HTML Help pages.  See the documentation for
+# a list of builtin themes.
+#
+html_theme = "furo"
+html_title = "Python Styleguide"
```
Comment (tomuben) [inline styleguide/conf.py:53]:
```suggestion
html_title = "Exasol Python Styleguide"
```

File: noxfile.py (hunk 2 above)
Comment (Nicoretti) [reply noxfile.py:18]: @tomuben thanks for clarifying. 
@tkilias , @tomuben if I have seen correctly, currently there is no "release" of this package yet, neither on `pipy` nor on `GH`. So personally I would rather wait for `sphinx-github-pages-generator` to provide package for the following reasonds:

* By initially having "pure" sphinx then using the generator we will validate that the results is the "same" (expected one)
* By having using it via the package, we can check if the package is correct and working
* It forces the project to do packaging (otherwise it increases the likelihood that packaging gets neglected in the long run)

File: pyproject.toml (hunk 16 above)
Comment (Nicoretti) [reply pyproject.toml]: good point have been unsure what and whom to add

File: styleguide/conf.py (hunk 17 above)
Comment (Nicoretti) [reply styleguide/conf.py:20]: Imho we should cut down on this over verbosity a styleguide which within the exasol domain/organization is obviosly
by exasol especially if it's the author etc.

Repeating namespaces etc. just increases bloat. :P

File: styleguide/conf.py (hunk 19 above)
Comment (Nicoretti) [reply styleguide/conf.py:53]: I would make the same argument (bloat etc.) as above

File: noxfile.py (hunk 2 above)
Comment (tkilias) [reply noxfile.py:18]: @Nicoretti you can see, making this release as your next task regarding the styleguide, ;-)

File: styleguide/drafts/tooling.rst [hunk 20]
```diff
@@ -241,6 +242,30 @@ Alternatives
 
 * ¯\\_(ツ)_/¯ code formatters themselves (black, yapf, ..)
 
+Pyupgrade_
+++++++++++
+`Project Homepage <Isort_www_>`_
+
+Helps migrating old code (constructs) to modern one.
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst:251]: General question, this is not only for python2 to python3, I guess. This is also between python3 versions

Comment (Nicoretti) [reply styleguide/drafts/tooling.rst:251]: No it is not only for `python2` to `python3`, it generally helps to update the code base to more modern constructs and api usages.

Example of `python 3.6` to `python 3.7`
```diff
-output = subprocess.run(['foo'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+output = subprocess.run(['foo'], capture_output=True)

-output = subprocess.run(['foo'], universal_newlines=True)
+output = subprocess.run(['foo'], text=True)
```

Comment (tkilias) [reply styleguide/drafts/tooling.rst:251]: Nice

File: styleguide/drafts/tooling.rst [hunk 21]
```diff
@@ -359,6 +361,72 @@ Alternatives
...
+* `rtd-theme <https://sphinx-rtd-theme.readthedocs.io/en/stable/>`_
+
+
+sphinx-github-pages-generator_
+++++++++++++++++++++++++++++++
+`Project Homepage <sphinx-github-pages-generator_www_>`_
+
+Once various versions of a project are published, different users of the library
+use different versions
+To provide a common look and feel between projects and between different versions of the project,
```
Comment (tkilias) [inline styleguide/drafts/tooling.rst]: is this a left over

**Review by tkilias — CHANGES_REQUESTED** [review]

File: styleguide/drafts/tooling.rst (hunk 21 above)
Comment (Nicoretti) [reply styleguide/drafts/tooling.rst]: yes thanks :D

**Review by tkilias — APPROVED** [review]
